synapse/storage/databases/main/keys.py

```diff
@@ -289,15 +288,19 @@ def _get_server_keys_json_txn(
                         "key_json",
                     ),
                 )
-                results[(server_name, key_id)] = [
-                    FetchKeyResultForRemote(
-                        # Cast to bytes since postgresql returns a memoryview.
-                        key_json=bytes(row["key_json"]),
-                        valid_until_ts=row["ts_valid_until_ms"],
-                        added_ts=row["ts_added_ms"],
-                    )
-                    for row in rows
-                ]
+
+                if not rows:
+                    continue
+
+                row = max(rows, key=lambda r: r["ts_added_ms"])
```

Comment:
Is this a meaningful change, is it possible for this query to return multiple rows?

Is it worth doing this selection on the DB? (ORDER BY ts_added_ms DESC LIMIT 1?)

I guess it doesn't make much difference: the biggest overhead is likely making `len(server_keys)` separate SELECTs?

Reply:
Oh, is the point that we were doing this logic in query_keys and we've just moved it into here?

Reply:
Yeah, exactly. It's non trivial to write a query to select the latest row per key ID, due to sqlite vs postgres.